File: waltid-libraries/credentials/waltid-digital-credentials/src/commonMain/kotlin/id/walt/credentials/display/CredentialTitles.kt
package id.walt.credentials.display

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Format-aware display title for a stored credential when OpenID4VCI card art is not used.
 */
object CredentialTitles {
    fun fromPayload(
        format: String,
        credentialDataJson: String?,
        displayName: String? = null,
        fallback: String? = null,
    ): String {
        displayName?.trim()?.takeIf { it.isNotBlank() }?.let { return it }
        val payload = credentialDataJson.toJsonObject()
        val extracted = when {
            format.isMdoc() -> mdocTitle(payload)
            format.isSdJwt() -> sdJwtTitle(payload)
            else -> w3cTitle(payload)
        }
        return extracted ?: fallback?.trim()?.takeIf { it.isNotBlank() } ?: format
    }

    private fun w3cTitle(payload: JsonObject?): String? {
        val types = payload.stringValues("type")
            ?: payload.obj("vc")?.stringValues("type")
            ?: return null
        return types
            .firstOrNull { !it.isGenericVerifiableCredentialType() }
            ?.let(::humanizeToken)
    }

    private fun sdJwtTitle(payload: JsonObject?): String? {
        val vct = payload.string("vct") ?: return null
        return humanizeToken(vct)
    }

    private fun mdocTitle(payload: JsonObject?): String? {
        val docType = payload.string("docType")
            ?: payload.string("doctype")
            ?: payload.string("doc_type")
            ?: return null
        return MdocFriendlyNames[docType]
            ?: MdocFriendlyNames[docType.substringBeforeLast('.')]
            ?: humanizeToken(docType)
    }

    private fun humanizeToken(raw: String): String? {
        val token = trailingToken(raw) ?: return null
        if (token.isGenericVerifiableCredentialType()) return null
        val words = token
            .replace(camelCaseBoundary, "$1 $2")
            .split(wordSeparators)
            .filter { it.isNotBlank() }
            .map { word -> word.lowercase().replaceFirstChar { it.titlecase() } }
        return words.joinToString(" ").takeIf { it.isNotBlank() }
    }

    private fun trailingToken(raw: String): String? {
        val value = raw.trim().takeIf { it.isNotBlank() } ?: return null
        val fromUrl = value.substringAfterLast('/').substringBefore('?').takeIf {
            value.contains("://") && it.isNotBlank()
        }
        val source = fromUrl ?: value
        val parts = source.split('/', '#', ':', '.').filter { it.isNotBlank() }
        val last = parts.lastOrNull() ?: return source
        return if (last.all(Char::isDigit) && parts.size >= 2) {
            "${parts[parts.lastIndex - 1]}_$last"
        } else {
            last
        }
    }

    private fun String.isMdoc(): Boolean {
        val normalized = lowercase()
        return normalized == "mso_mdoc" || normalized == "mdoc" || normalized.contains("mdoc")
    }

    private fun String.isSdJwt(): Boolean {
        val normalized = lowercase()
        return normalized.contains("sd-jwt") || normalized.contains("sd_jwt") || normalized == "dc+sd-jwt" || normalized == "vc+sd-jwt"
    }

    private fun String.isGenericVerifiableCredentialType(): Boolean =
        trailingToken(this)?.equals("VerifiableCredential", ignoreCase = true) == true

    private fun String?.toJsonObject(): JsonObject? =
        this?.trim()?.takeIf { it.isNotBlank() }?.let { raw ->
            runCatching { Json.parseToJsonElement(raw).jsonObject }.getOrNull()
        }

    private fun JsonObject?.string(key: String): String? =
        this?.get(key)?.let { element ->
            (element as? JsonPrimitive)?.contentOrNull?.trim()?.takeIf { it.isNotBlank() }
        }

    private fun JsonObject?.obj(key: String): JsonObject? =
        this?.get(key) as? JsonObject

    private fun JsonObject?.stringValues(key: String): List<String>? {
        val element = this?.get(key) ?: return null
        return when (element) {
            is JsonArray -> element.mapNotNull { (it as? JsonPrimitive)?.contentOrNull?.trim()?.takeIf { value -> value.isNotBlank() } }
                .takeIf { it.isNotEmpty() }
            is JsonPrimitive -> element.contentOrNull?.trim()?.takeIf { it.isNotBlank() }?.let { listOf(it) }
            else -> null
        }
    }

    private val wordSeparators = Regex("[_\\-. ]+")
    private val camelCaseBoundary = Regex("([a-z])([A-Z])")

    internal val MdocFriendlyNames = mapOf(
        "org.iso.18013.5.1.mDL" to "Mobile Driving Licence",
        "org.iso.18013.5.1" to "Mobile Driving Licence",
        "eu.europa.ec.eudi.pid.1" to "PID",
        "eu.europa.ec.eudi.pid" to "PID",
        "org.iso.23220.photoid.1" to "Photo ID",
        "org.iso.23220.1" to "Photo ID",
        "eu.europa.ec.eudi.mdl.1" to "Mobile Driving Licence",
    )
}

File: waltid-libraries/protocols/waltid-wallet-sdk-ios/Sources/WalletSDK/CredentialTitles.swift
import Foundation

/// Format-aware display title for a stored credential when OpenID4VCI card art is not used.
public enum CredentialTitles {
    /// Resolves a human-readable credential title from stored display metadata or payload fields.
    ///
    /// Prefers an explicit display name, then a format-specific title extracted from the credential
    /// JSON, then `fallback`, and finally the raw format string.
    ///
    /// - Parameters:
    ///   - format: Credential format used to choose the payload title rules.
    ///   - credentialDataJSON: Serialized credential payload, when one is available.
    ///   - displayName: Issuer or wallet display name that should win over extracted titles.
    ///   - fallback: Label to use when neither display metadata nor the payload yields a title.
    public static func displayName(
        format: String,
        credentialDataJSON: String?,
        displayName: String? = nil,
        fallback: String? = nil
    ) -> String {
        if let displayName, !displayName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return displayName.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        let payload = credentialDataJSON.flatMap(parseJSON)
        let extracted: String?
        if isMdoc(format) {
            extracted = mdocTitle(payload)
        } else if isSdJwt(format) {
            extracted = sdJwtTitle(payload)
        } else {
            extracted = w3cTitle(payload)
        }
        return extracted
            ?? fallback?.trimmingCharacters(in: .whitespacesAndNewlines).nonEmpty
            ?? format
    }

    private static func w3cTitle(_ payload: [String: Any]?) -> String? {
        let types = stringValues(payload?["type"]) ?? stringValues((payload?["vc"] as? [String: Any])?["type"])
        return types?
            .first { !isGenericVerifiableCredentialType($0) }
            .flatMap(humanizeToken)
    }

    private static func sdJwtTitle(_ payload: [String: Any]?) -> String? {
        guard let vct = payload?["vct"] as? String else { return nil }
        return humanizeToken(vct)
    }

    private static func mdocTitle(_ payload: [String: Any]?) -> String? {
        let docType = (payload?["docType"] as? String)
            ?? (payload?["doctype"] as? String)
            ?? (payload?["doc_type"] as? String)
        guard let docType else { return nil }
        if let mapped = mdocFriendlyNames[docType] { return mapped }
        let namespace = docType.split(separator: ".").dropLast().joined(separator: ".")
        return mdocFriendlyNames[namespace] ?? humanizeToken(docType)
    }

    private static func humanizeToken(_ raw: String) -> String? {
        guard let token = trailingToken(raw), !isGenericVerifiableCredentialType(token) else { return nil }
        let spaced = token.replacingOccurrences(
            of: "([a-z])([A-Z])",
            with: "$1 $2",
            options: .regularExpression
        )
        let words = spaced
            .split { $0 == "_" || $0 == "-" || $0 == "." || $0 == " " }
            .filter { !$0.isEmpty }
            .map { word in
                word.lowercased().prefix(1).uppercased() + word.lowercased().dropFirst()
            }
        let joined = words.joined(separator: " ")
        return joined.isEmpty ? nil : joined
    }

    private static func trailingToken(_ raw: String) -> String? {
        let value = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty else { return nil }
        let source: String
        if value.contains("://") {
            source = value.split(separator: "/").last.map(String.init)?.split(separator: "?").first.map(String.init) ?? value
        } else {
            source = value
        }
        let parts = source.split { $0 == "/" || $0 == "#" || $0 == ":" || $0 == "." }.map(String.init).filter { !$0.isEmpty }
        guard let last = parts.last else { return source }
        if last.allSatisfy(\.isNumber), parts.count >= 2 {
            return "\(parts[parts.count - 2])_\(last)"
        }
        return last
    }

    private static func isMdoc(_ format: String) -> Bool {
        let normalized = format.lowercased()
        return normalized == "mso_mdoc" || normalized == "mdoc" || normalized.contains("mdoc")
    }

    private static func isSdJwt(_ format: String) -> Bool {
        let normalized = format.lowercased()
        return normalized.contains("sd-jwt") || normalized.contains("sd_jwt")
    }

    private static func isGenericVerifiableCredentialType(_ value: String) -> Bool {
        trailingToken(value)?.caseInsensitiveCompare("VerifiableCredential") == .orderedSame
    }

    private static func parseJSON(_ raw: String) -> [String: Any]? {
        guard let data = raw.data(using: .utf8) else { return nil }
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }

    private static func stringValues(_ value: Any?) -> [String]? {
        switch value {
        case let values as [String]:
            return values.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
        case let value as String:
            let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
            return trimmed.isEmpty ? nil : [trimmed]
        default:
            return nil
        }
    }

    private static let mdocFriendlyNames = [
        "org.iso.18013.5.1.mDL": "Mobile Driving Licence",
        "org.iso.18013.5.1": "Mobile Driving Licence",
        "eu.europa.ec.eudi.pid.1": "PID",
        "eu.europa.ec.eudi.pid": "PID",
        "org.iso.23220.photoid.1": "Photo ID",
        "org.iso.23220.1": "Photo ID",
        "eu.europa.ec.eudi.mdl.1": "Mobile Driving Licence",
    ]
}

private extension String {
    var nonEmpty: String? { isEmpty ? nil : self }
}
